### PR TITLE
Add the `Symfony\Config\config()` function

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

This PR adds a new `config()` function to configure a Symfony application:

```php
<?php

use function Symfony\Config\config;

return config(
    framework: [
        'cache' => [
            'prefix_seed' => 'seed_'.$this->env.'_',
        ],
    ],
    security: [
        'access_decision_manager' => [
            'strategy' => 'unanimous',
        ],
    ]
);
```

The fonction is generated at cache warmup, allowing to generate the corresponding phpdoc along with Jetbrain's `ArrayShape` attribute. Here a sneak pick of what the function looks like:

```php
<?php

namespace Symfony\Config;

use Jetbrains\PhpStorm\ArrayShape;

/**
 * This class is automatically generated to help in creating a config.
 *
 * @param array{
 *     secret?: string|int|float|bool,
 *     http_method_override?: bool,
 *     trust_x_sendfile_type_header?: string|int|float|bool,
 *     ide?: string|int|float|bool,
 *     test?: bool,
 *     ...
 *     form?: array{
 *         enabled?: bool,
 *         csrf_protection?: array{
 *             enabled?: string|int|float|bool,
 *             token_id?: string|int|float|bool,
 *             field_name?: string|int|float|bool,
 *             field_attr?: array<array-key, mixed>,
 *         },
 *     },
 *     ...
 *   } $framework
 *
 * @param array{
 *     access_denied_url?: string|int|float|bool,
 *     session_fixation_strategy?: "none"|"migrate"|"invalidate",
 *     hide_user_not_found?: bool,
 *     erase_credentials?: bool,
 *     access_decision_manager?: array{
 *         strategy?: "affirmative"|"consensus"|"unanimous"|"priority",
 *         service?: string|int|float|bool,
 *         strategy_service?: string|int|float|bool,
 *         allow_if_all_abstain?: bool,
 *         allow_if_equal_granted_denied?: bool,
 *     },
 *     ...
 *   } $security
 */
function config(
    // framework:
    #[ArrayShape([
        'secret' => 'string|int|float|bool',
        'http_method_override' => 'bool', /* Set true to enable support for the '_method' request parameter to determine the intended HTTP method on POST requests. Note: When using the HttpCache, you need to call the method in your front controller instead. Default value: false. */
        'trust_x_sendfile_type_header' => 'string|int|float|bool', /* Set true to enable support for xsendfile in binary file responses. Default value: "%env(bool:default::SYMFONY_TRUST_X_SENDFILE_TYPE_HEADER)%". */
        'ide' => 'string|int|float|bool', /* Default value: "%env(default::SYMFONY_IDE)%". */
        'test' => 'bool',
        // ...
        'form' => [
            'enabled' => 'bool', /* Default value: false. */
            'csrf_protection' => [
                'enabled' => 'string|int|float|bool', /* Default value: null. */
                'token_id' => 'string|int|float|bool', /* Default value: null. */
                'field_name' => 'string|int|float|bool', /* Default value: "_token". */
                'field_attr' => 'array<array-key, mixed>',
            ],
    ], /* Form configuration */
    // ...
    ])] array $framework = [],
    // security:
    #[ArrayShape([
        'access_denied_url' => 'string|int|float|bool', /* Default value: null. */
        'session_fixation_strategy' => '"none"|"migrate"|"invalidate"', /* Default value: "migrate". */
        'hide_user_not_found' => 'bool', /* Default value: true. */
        'erase_credentials' => 'bool', /* Default value: true. */
        'access_decision_manager' => [
            'strategy' => '"affirmative"|"consensus"|"unanimous"|"priority"',
            'service' => 'string|int|float|bool',
            'strategy_service' => 'string|int|float|bool',
            'allow_if_all_abstain' => 'bool', /* Default value: false. */
            'allow_if_equal_granted_denied' => 'bool', /* Default value: true. */
        ],
        // ...
    ])] array $security = []): \Closure {
    // ...
}
```

### CODE REVIEW

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ *
+ * @internal
+ */
+final class ArrayShapeGenerator
+{
+    public const FORMAT_PHPDOC = 'phpdoc';
+    public const FORMAT_JETBRAINS_ARRAY_SHAPE = 'jetbrains_array_shape';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
    public const FORMAT_JETBRAINS_ATTRIBUTE = 'jetbrains_attribute';
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        if (self::FORMAT_JETBRAINS_ARRAY_SHAPE === $format) {
+            return static::doGenerateJetBrainsArrayShape($node);
+        }
+
+        throw new \LogicException(sprintf('Unsupported format to generate array shape. Expected one of "%s", got "%s".', implode('", "', [self::FORMAT_PHPDOC, self::FORMAT_JETBRAINS_ARRAY_SHAPE]), $format));
+    }
+
+    private static function doGeneratePhpDoc(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if ($node instanceof ArrayNode) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
        if (!$node instanceof ArrayNode) { return
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            $children = $node->getChildren();
+            $arrayShape = 'array';
+
+            if (!$children) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
            $arrayShape = 'array';

            if (!$children = $node->getChildren()) {
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+            return $arrayShape;
+        }
+
+        return $node->getName();
+    }
+
+    private static function doGenerateJetBrainsArrayShape(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if ($node instanceof ArrayNode) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
        if (!$node instanceof ArrayNode) {return
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            $children = $node->getChildren();
+
+            $shape = '';
+            if (1 === $nestingLevel) {
+                $shape = '#[ArrayShape(';
+            } else {
+                if (!$children) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
            } elseif (!$children) {
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        if ($node instanceof IntegerNode) {
+            $type = 'int<%s, %s>';
+        } elseif ($node instanceof FloatNode) {
+            $type = 'float<%s, %s>';
+        } else {
+            $type = 'int<%s, %s>|float<%1$s, %2$s>';
+        }
+
+        $min = null !== $node->getMin() ? $node->getMin() : 'min';
+        $max = null !== $node->getMax() ? $node->getMax() : 'max';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
        $min = $node->getMin() ?? 'min';
        $max = $node->getMax() ?? 'max';
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+        return \sprintf($type, $min, $max);
+    }
+
+    private static function prependPhpDocWithStar(string $shape): string
+    {
+        return strtr($shape, ["\n" => "\n * "]);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]: str_replace

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    private static function generateInlinePhpDocForNode(NodeInterface $node): string
+    {
+        $hasContent = false;
+        $comment = ' /* ';
+
+        if ($node->hasDefaultValue() || $node->getInfo() !== null || $node->isDeprecated()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
        if ($node->hasDefaultValue() || $node->getInfo() || $node->isDeprecated()) {
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            }
+
+            if ($info = $node->getInfo()) {
+                $hasContent = true;
+                $comment .= $info.' ';
+            }
+
+            if ($node->hasDefaultValue() && !\is_array($defaultValue = $node->getDefaultValue())) {
+                $hasContent = true;
+                $comment .= 'Default value: '.json_encode($defaultValue).'. ';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
                $comment .= 'Default: '.json_encode($defaultValue).'. ';
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+ */
+final class ArrayShapeGenerator
+{
+    public const FORMAT_PHPDOC = 'phpdoc';
+    public const FORMAT_JETBRAINS_ARRAY_SHAPE = 'jetbrains_array_shape';
+
+    /**
+     * @param self::FORMAT_* $format
+     */
+    public static function generate(ArrayNode $node, string $format = self::FORMAT_PHPDOC): string
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]: I wouldn't add a default value, it makes things less explicit

File: src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypesConfig.php [hunk 11]
```diff
@@ -128,6 +130,46 @@ public function nested(array $value = []): \Symfony\Config\ScalarNormalizedTypes
...
+     *     keyed_list_object?: array<array-key, mixed>,
+     *     nested?: array{
+     *         nested_object?: array{
+     *             enabled?: bool,
+     *         },
+     *         nested_list_object?: array<array-key, mixed>,
+     *     },
+     * } $config
+     */
+    public function configure(#[ArrayShape([
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypesConfig.php]: let's ensure each argument start on its own line with a starting comment to help the reader:
```suggestion
    public function configure(
        // config:
        #[ArrayShape([
```

File: src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php [hunk 12]
```diff
@@ -36,13 +37,15 @@
 class PhpFileLoader extends FileLoader
 {
     protected bool $autoRegisterAliasesForSinglyImplementedInterfaces = false;
+    private static array $configurations = [];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php]: why static? looks like a bad idea to me :)

File: src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php [hunk 13]
```diff
@@ -36,13 +37,15 @@
...
 
     public function __construct(
         ContainerBuilder $container,
         FileLocatorInterface $locator,
         ?string $env = null,
         private ?ConfigBuilderGeneratorInterface $generator = null,
         bool $prepend = false,
+        private ?ConfigFunctionGenerator $configFunctionGenerator = null,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php]: could we find a way that doesn't require changing the constructor? can't we find a way to use the ConfigBuilderGeneratorInterface instance?

File: src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+final class ConfigFunctionGenerator
+{
+    public function __construct(
+        private string $outputDir,
+    ) {
+    }
+
+    /**
+     * @param array<string, ConfigurationInterface> $configurations Configurations indexed by their alias.
+     * @return \Closure
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php]:
```suggestion
```

File: src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+ */
+final class ConfigFunctionGenerator
+{
+    public function __construct(
+        private string $outputDir,
+    ) {
+    }
+
+    /**
+     * @param array<string, ConfigurationInterface> $configurations Configurations indexed by their alias.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php]:
```suggestion
     * @param array<string, ConfigurationInterface> $configurations Configurations indexed by their alias
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+                }
+            }
+
+            $shape .= '['.PHP_EOL;
+
+            /** @var NodeInterface $child */
+            foreach ($children as $child) {
+                $shape .= \sprintf("%s'%s' => ", str_repeat(' ', $nestingLevel*4), $child->getName());
+                if ($child instanceof ArrayNode) {
+                    $shape .= static::doGenerateJetBrainsArrayShape($child, $nestingLevel+1);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
                    $shape .= static::doGenerateJetBrainsArrayShape($child, 1 + $nestingLevel);
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+                if (!$children) {
+                    return "'array<array-key, mixed>'";
+                }
+            }
+
+            $shape .= '['.PHP_EOL;
+
+            /** @var NodeInterface $child */
+            foreach ($children as $child) {
+                $shape .= \sprintf("%s'%s' => ", str_repeat(' ', $nestingLevel*4), $child->getName());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
                $shape .= \sprintf("%s'%s' => ", str_repeat(' ', 4 * $nestingLevel), $child->getName());
```

File: src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+                if ($child instanceof ArrayNode) {
+                    $shape .= static::doGenerateJetBrainsArrayShape($child, $nestingLevel+1);
+                } else {
+                    $shape .= "'".static::handleNodeType($child)."'";
+                }
+
+                $shape .= ','.static::generateInlinePhpDocForNode($child).PHP_EOL;
+            }
+
+            $shape .= str_repeat(' ', ($nestingLevel-1)*4).']';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php]:
```suggestion
            $shape .= str_repeat(' ', 4 * ($nestingLevel - 1)).']';
```

**Review by nicolas-grekas — COMMENTED** [review]
First quick review

**Comment by alexandre-daubois** [comment]
@nicolas-grekas thanks for the review, I addressed all comments. I found a way to keep PhpFileLoader constructor as-is

**Comment by alexandre-daubois** [comment]
Tests added to cover the `config()` function generation

**Comment by alexandre-daubois** [comment]
https://github.com/symfony/symfony/pull/58771

**Pull request closed without merging**